### PR TITLE
feat: create assets via postDeploymentSetup only

### LINKED ISSUE #498: Create assets in post deployment script
Create jobs and collections in post deployment. Pick up from https://github.com/razor-network/datasources

### PR DESCRIPTION
**Pull request by abhishekvispute**
fixes #498 #518

### CODE REVIEW

File: migrations/migrationHelpers.js [hunk 1]
```diff
@@ -90,7 +91,11 @@ const deployContract = async (
     config.contract = 'contracts/tokenization/RAZOR.sol:RAZOR';
   }
 
-  await hre.run('verify:verify', config);
+  try {
+    await hre.run('verify:verify', config);
+  } catch (err) {
+    console.error('Etherscan verification failed');
```
Comment (SkandaBhat) [inline migrations/migrationHelpers.js:97]: Log the error as well here.

File: migrations/migrationHelpers.js [hunk 2]
```diff
@@ -117,11 +122,40 @@ const getdeployedContractInstance = async (
   return { Contract, contractInstance };
 };
 
+const source = 'https://raw.githubusercontent.com/razor-network/datasources/master';
```
Comment (SkandaBhat) [inline migrations/migrationHelpers.js:125]: String consts need to be in uppercase. const SOURCE = '...'

File: migrations/migrationHelpers.js [hunk 3]
```diff
@@ -117,11 +122,40 @@ const getdeployedContractInstance = async (
...
+
+const getJobs = async () => {
+  try {
+    const jobs = await axios.get(`${source}/jobs.json`);
+    return jobs.data;
+  } catch (error) {
+    console.log(error.response.body);
```
Comment (SkandaBhat) [inline migrations/migrationHelpers.js:132]: Be more descriptive here. `console.log("Error while fetching jobs: ", error.response.body);`

File: migrations/migrationHelpers.js [hunk 4]
```diff
@@ -117,11 +122,40 @@ const getdeployedContractInstance = async (
...
+    return null;
+  }
+};
+
+const getCollections = async () => {
+  try {
+    const collections = await axios.get(`${source}/collections.json`);
+    return collections.data;
+  } catch (error) {
+    console.log(error.response.body);
```
Comment (SkandaBhat) [inline migrations/migrationHelpers.js:142]: Same as above.

File: migrations/src/postDeploymentSetup.js [hunk 5]
```diff
@@ -127,12 +131,38 @@ module.exports = async () => {
...
+
+  console.log('Creating Collections');
+  console.log('Waiting for Confirm state : 4.......');
+  let state = 0;
+  const numStates = await stakeManager.NUM_STATES();
+  const stateLength = (BigNumber.from(await stakeManager.epochLength())).div(numStates);
+
+  while (state !== 4) {
+    const blockNumber = await ethers.provider.getBlockNumber();
+    state = Number(((BigNumber.from(blockNumber)).div(stateLength)).mod(numStates));
```
Comment (SkandaBhat) [inline migrations/src/postDeploymentSetup.js:157]: Calculate state using a utility function.

File: migrations/src/postDeploymentSetup.js [hunk 6]
```diff
@@ -127,12 +131,38 @@ module.exports = async () => {
...
+
+  for (let i = 0; i < jobs.length; i++) {
+    const job = jobs[i];
+    await assetManager.createJob(job.weight, job.power, job.selectorType, job.name, job.selector, job.url);
+    console.log(`Job Created :  ${job.name}`);
+  }
+
+  console.log('Creating Collections');
+  console.log('Waiting for Confirm state : 4.......');
+  let state = 0;
```
Comment (SkandaBhat) [inline migrations/src/postDeploymentSetup.js:151]: state should be initialised to current state and not zero. If it is actually confirm state, then we are waiting for 10 seconds for no reason.

File: migrations/src/postDeploymentSetup.js [hunk 7]
```diff
@@ -127,12 +131,38 @@ module.exports = async () => {
...
+  const numStates = await stakeManager.NUM_STATES();
+  const stateLength = (BigNumber.from(await stakeManager.epochLength())).div(numStates);
+
+  while (state !== 4) {
```
Comment (SkandaBhat) [inline migrations/src/postDeploymentSetup.js:155]: This entire thing can be a utility function for added readability. `await waitForConfirmState();`

File: migrations/src/postDeploymentSetup.js [hunk 8]
```diff
@@ -127,12 +131,38 @@ module.exports = async () => {
...
+    const blockNumber = await ethers.provider.getBlockNumber();
+    state = Number(((BigNumber.from(blockNumber)).div(stateLength)).mod(numStates));
+    console.log('Current state', state);
+    await sleep(10000);
+  }
+
+  for (let i = 0; i < collections.length; i++) {
+    const collection = collections[i];
+    await assetManager.createCollection(collection.jobIDs, collection.aggregationMethod, collection.power, collection.name);
```
Comment (SkandaBhat) [inline migrations/src/postDeploymentSetup.js:164]: Should check for confirm state before creating each collection, and wait for confirm state if not in confirm state.

**Review by SkandaBhat — CHANGES_REQUESTED** [review]

File: migrations/src/postDeploymentSetup.js (hunk 8 above)
Comment (abhishekvispute) [reply migrations/src/postDeploymentSetup.js:157]: resolved https://github.com/razor-network/contracts/pull/579/commits/57925d7b53c6d85e2116b416953ea4d30159789e

File: migrations/src/postDeploymentSetup.js (hunk 7 above)
Comment (abhishekvispute) [reply migrations/src/postDeploymentSetup.js]: resolved 57925d7

File: migrations/src/postDeploymentSetup.js (hunk 6 above)
Comment (abhishekvispute) [reply migrations/src/postDeploymentSetup.js]: resolved 57925d7

File: migrations/src/postDeploymentSetup.js (hunk 5 above)
Comment (abhishekvispute) [reply migrations/src/postDeploymentSetup.js]: resolved 57925d7

File: migrations/migrationHelpers.js (hunk 4 above)
Comment (abhishekvispute) [reply migrations/migrationHelpers.js]: resolved 57925d7

File: migrations/migrationHelpers.js (hunk 3 above)
Comment (abhishekvispute) [reply migrations/migrationHelpers.js]: resolved 57925d7

File: migrations/migrationHelpers.js (hunk 2 above)
Comment (abhishekvispute) [reply migrations/migrationHelpers.js]: resolved 57925d7

File: migrations/migrationHelpers.js (hunk 1 above)
Comment (abhishekvispute) [reply migrations/migrationHelpers.js]: resolved https://github.com/razor-network/contracts/pull/579/commits/323a53f4abccbe76f5a7a48631eed413176bec18

**Comment by abhishekvispute** [comment]
This is how it looks now
```
Waiting for post-deployment setup transactions to get confirmed
Creating Jobs
Job Created :  ethusd_gemini
Job Created :  ethusd_kucoin
Job Created :  ethusd_kraken
Job Created :  btcusd_kucoin
Job Created :  btcusd_gemini
Job Created :  btcusd_kraken
Creating Collections
Waiting for Confirm state : 4.......
Current state 2
Current state 2
Current state 2
Current state 3
Current state 3
Current state 3
Current state 3
Current state 3
Current state 3
Current state 3
Current state 3
Current state 3
Current state 3
Current state 3
Current state 3
Current state 4
Collection Created :  ethCollectionMean
Collection Created :  btcCollectionMean
Contracts deployed successfully & initial setup is done
Done
```

**Review by SkandaBhat — APPROVED** [review]

**Pull request merged**